File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_diameter = 0
        def calculate_depth(node: Optional[TreeNode]) -> int:
            if not node:
                return 0

            left_depth = calculate_depth(node.left)
            right_depth = calculate_depth(node.right)
            current_diameter = left_depth + right_depth
            self.max_diameter = max(self.max_diameter , current_diameter)
            return max(left_depth, right_depth) + 1
        calculate_depth(root)
        return self.max_diameter
        

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return dfs(root, root.val);
    }

    private int dfs(TreeNode node, int maxSoFar) {
        if (node == null ) {
            return 0;
        }
        int good = (node.val >= maxSoFar) ? 1 : 0;
        int newMax = Math.max(node.val, maxSoFar);
        good += dfs(node.left, newMax);
        good += dfs(node.right, newMax);
        return good;
    }
}
//     private int dfs(TreeNode node, int maxSoFar) {
//         // Base case: null nodes don't contribute to the count
//         if (node == null) {
//             return 0;
//         }

//         // Check if the current node is a "good" node
//         int good = (node.val >= maxSoFar) ? 1 : 0;
        
//         // Update the max value for the path going forward
//         int newMax = Math.max(node.val, maxSoFar);
        
//         // Recursively count good nodes in left and right subtrees
//         good += dfs(node.left, newMax);
//         good += dfs(node.right, newMax);
        
//         return good;
//     }
// }
